### PR TITLE
GitHub Actions CI Overhaul

### PR DESCRIPTION
**Pull request by achristoforides**
### PR Description

This PR introduces changes to our CMake configuration files in addition to changes to our GitHub Actions workflows. _**It additionally introduces a version upgrade for GoogleTest and GoogleBenchmark as the version we were using did not have a fix which was causing errors when using the sanitizers.**_

### Updates to GitHub Actions Workflows
The existing BuildAndTest workflow was modified to no longer run the TSAN build in addition to now making the action auto-run when a push to master occurs.

A new workflow (SanitizerChecks) was added which now runs 3 jobs: a build with ASAN (with leak checking turned on), a build with TSAN, and a build with UBSAN. MSAN was not added as its usage necessitates that all libraries used in the software (including the implementations of the standard libraries) must be instrumented with MSAN.

Note: The required infrastructure is in place for allowing MSAN to be used. If somebody turns on the flag which enables it, they will receive an error stating that the infrastructure is in place but it is not able to be used yet.

### GoogleTest, GoogleMock, and GoogleBenchmark

All versions of the above have been updated. Additionally, GoogleBenchmark was converted to be a git submodule. The reason for these updates is that googletest and googlemock had issues with some of the sanitizers that have been fixed in recent versions.

Our previous version of GoogleBenchmark had issues when building benchmark, gtest, and gmock as shared libraries. The new version fixes this.

### CMake Changes

Lastly, the CMake files were updated to allow for enabling and disabling of the aforementioned sanitizers. Additionally, a few unrelated changes were made in those files to clean up code which is no longer necessary and formatting changes. A notable change however is the removal of the CMake code which always forced CppMicroServices to link against `libc++` and `libc++abi` when building with Clang on Linux; this is no longer needed.

Another major change included in the CMake files is building `googletest`, `googlemock`, and `benchmark` as shared libraries and subsequently linking them into test executables dynamically. With the sanitizers turned on, statically linking them caused issues and I additionally do not see a reason for doing this anymore.

Note: Any changes in cpp/hpp/h files are to fix warnings and errors that the sanitizers identified.

### CODE REVIEW

File: tsan_suppressions.txt [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # ServiceTracker/ServiceListener races
 race:cppmicroservices::ServiceListeners::RemoveServiceListener
 race:cppmicroservices::ServiceTracker*
+race:ServiceTracker.tpp
```
Comment (jeffdiclemente) [inline tsan_suppressions.txt:4]: do you have the tsan output for this race? I'm interested to see what it is.

File: valgrind_suppressions.txt [hunk 2]
```diff
@@ -416,3 +416,20 @@
...
+   fun:_ZN7testing8internal20PrintBytesInObjectToEPKhmPSo
+   ...
+}
+
+{
+   <usDeclarativeServicesTestsSuppression3>
+   Memcheck:Cond
+   ...
+   fun:_ZN7testing8internal20PrintBytesInObjectToEPKhmPSo
+   ...
```
Comment (jeffdiclemente) [inline valgrind_suppressions.txt:433]: which tests were these valgrind errors coming from?
If I remember correctly, these come from not having an overloaded << operator for a user-defined type.

**Review by jeffdiclemente — CHANGES_REQUESTED** [review]

File: tsan_suppressions.txt (hunk 1 above)
Comment (achristoforides) [reply tsan_suppressions.txt:4]: https://github.com/CppMicroServices/CppMicroServices/runs/3275718509

It looks like adding ServiceTracker.tpp is not necessary though (after verifying that I do not see it locally without this change). I will remove it.

File: valgrind_suppressions.txt (hunk 2 above)
Comment (achristoforides) [reply valgrind_suppressions.txt:433]: The errors that these suppress come from gtest/gmock's warmup for `usDeclarativeServicesTests`. Looks like it has something to do with `ComponentMetadata`? Should I declare an overloaded << operator for it? Additionally the error only shows up a couple times in the log but the error summary for valgrind states there are 2448 errors in total; including the suppression above removes them all (just for context).

```
==32383== Memcheck, a memory error detector
==32383== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==32383== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==32383== Command: ./build/bin/usDeclarativeServicesTests
==32383==
==32383== Use of uninitialised value of size 8
==32383==    at 0x7471911: _itoa_word (_itoa.c:180)
==32383==    by 0x747500D: vfprintf (vfprintf.c:1642)
==32383==    by 0x74A19DF: vsnprintf (vsnprintf.c:114)
==32383==    by 0x747E0CE: snprintf (snprintf.c:33)
==32383==    by 0x64C4D2D: testing::internal::PrintBytesInObjectTo(unsigned char const*, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x63ADBB: void testing::internal::PrintTo<cppmicroservices::scrimpl::metadata::ComponentMetadata>(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&, std::ostream*) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x63A987: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > testing::PrintToString<std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> >(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x639F5E: testing::internal::ParameterizedTestSuiteInfo<cppmicroservices::scrimpl::ComponentManagerImplParameterizedTest>::RegisterTests() (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x64B6E68: testing::internal::UnitTestImpl::PostFlagParsingInit() (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64CD7D8: void testing::internal::InitGoogleTestImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x6067208: void testing::internal::InitGoogleMockImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgmock.so.1.11.0)
==32383==    by 0x7B373F: main (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==
==32383== Conditional jump or move depends on uninitialised value(s)
==32383==    at 0x7471918: _itoa_word (_itoa.c:180)
==32383==    by 0x747500D: vfprintf (vfprintf.c:1642)
==32383==    by 0x74A19DF: vsnprintf (vsnprintf.c:114)
==32383==    by 0x747E0CE: snprintf (snprintf.c:33)
==32383==    by 0x64C4D2D: testing::internal::PrintBytesInObjectTo(unsigned char const*, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x63ADBB: void testing::internal::PrintTo<cppmicroservices::scrimpl::metadata::ComponentMetadata>(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&, std::ostream*) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x63A987: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > testing::PrintToString<std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> >(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x639F5E: testing::internal::ParameterizedTestSuiteInfo<cppmicroservices::scrimpl::ComponentManagerImplParameterizedTest>::RegisterTests() (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x64B6E68: testing::internal::UnitTestImpl::PostFlagParsingInit() (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64CD7D8: void testing::internal::InitGoogleTestImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x6067208: void testing::internal::InitGoogleMockImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgmock.so.1.11.0)
==32383==    by 0x7B373F: main (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==
==32383== Conditional jump or move depends on uninitialised value(s)
==32383==    at 0x7475114: vfprintf (vfprintf.c:1642)
==32383==    by 0x74A19DF: vsnprintf (vsnprintf.c:114)
==32383==    by 0x747E0CE: snprintf (snprintf.c:33)
==32383==    by 0x64C4D2D: testing::internal::PrintBytesInObjectTo(unsigned char const*, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x63ADBB: void testing::internal::PrintTo<cppmicroservices::scrimpl::metadata::ComponentMetadata>(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&, std::ostream*) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x63A987: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > testing::PrintToString<std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> >(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x639F5E: testing::internal::ParameterizedTestSuiteInfo<cppmicroservices::scrimpl::ComponentManagerImplParameterizedTest>::RegisterTests() (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x64B6E68: testing::internal::UnitTestImpl::PostFlagParsingInit() (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64CD7D8: void testing::internal::InitGoogleTestImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x6067208: void testing::internal::InitGoogleMockImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgmock.so.1.11.0)
==32383==    by 0x7B373F: main (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==
==32383== Conditional jump or move depends on uninitialised value(s)
==32383==    at 0x7475C4C: vfprintf (vfprintf.c:1642)
==32383==    by 0x74A19DF: vsnprintf (vsnprintf.c:114)
==32383==    by 0x747E0CE: snprintf (snprintf.c:33)
==32383==    by 0x64C4D2D: testing::internal::PrintBytesInObjectTo(unsigned char const*, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x63ADBB: void testing::internal::PrintTo<cppmicroservices::scrimpl::metadata::ComponentMetadata>(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&, std::ostream*) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x63A987: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > testing::PrintToString<std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> >(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x639F5E: testing::internal::ParameterizedTestSuiteInfo<cppmicroservices::scrimpl::ComponentManagerImplParameterizedTest>::RegisterTests() (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x64B6E68: testing::internal::UnitTestImpl::PostFlagParsingInit() (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64CD7D8: void testing::internal::InitGoogleTestImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x6067208: void testing::internal::InitGoogleMockImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgmock.so.1.11.0)
==32383==    by 0x7B373F: main (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==
==32383== Conditional jump or move depends on uninitialised value(s)
==32383==    at 0x7475114: vfprintf (vfprintf.c:1642)
==32383==    by 0x74A19DF: vsnprintf (vsnprintf.c:114)
==32383==    by 0x747E0CE: snprintf (snprintf.c:33)
==32383==    by 0x64C76C1: testing::(anonymous namespace)::PrintByteSegmentInObjectTo(unsigned char const*, unsigned long, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64C4D91: testing::internal::PrintBytesInObjectTo(unsigned char const*, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x63ADBB: void testing::internal::PrintTo<cppmicroservices::scrimpl::metadata::ComponentMetadata>(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&, std::ostream*) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x63A987: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > testing::PrintToString<std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> >(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x639F5E: testing::internal::ParameterizedTestSuiteInfo<cppmicroservices::scrimpl::ComponentManagerImplParameterizedTest>::RegisterTests() (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x64B6E68: testing::internal::UnitTestImpl::PostFlagParsingInit() (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64CD7D8: void testing::internal::InitGoogleTestImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x6067208: void testing::internal::InitGoogleMockImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgmock.so.1.11.0)
==32383==    by 0x7B373F: main (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==
==32383== Conditional jump or move depends on uninitialised value(s)
==32383==    at 0x7475C4C: vfprintf (vfprintf.c:1642)
==32383==    by 0x74A19DF: vsnprintf (vsnprintf.c:114)
==32383==    by 0x747E0CE: snprintf (snprintf.c:33)
==32383==    by 0x64C76C1: testing::(anonymous namespace)::PrintByteSegmentInObjectTo(unsigned char const*, unsigned long, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64C4D91: testing::internal::PrintBytesInObjectTo(unsigned char const*, unsigned long, std::ostream*) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x63ADBB: void testing::internal::PrintTo<cppmicroservices::scrimpl::metadata::ComponentMetadata>(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&, std::ostream*) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x63A987: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > testing::PrintToString<std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> >(std::shared_ptr<cppmicroservices::scrimpl::metadata::ComponentMetadata> const&) (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x639F5E: testing::internal::ParameterizedTestSuiteInfo<cppmicroservices::scrimpl::ComponentManagerImplParameterizedTest>::RegisterTests() (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==    by 0x64B6E68: testing::internal::UnitTestImpl::PostFlagParsingInit() (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x64CD7D8: void testing::internal::InitGoogleTestImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgtest.so.1.11.0)
==32383==    by 0x6067208: void testing::internal::InitGoogleMockImpl<char>(int*, char**) (in /home/vagrant/CppMicroServices/build/lib/libgmock.so.1.11.0)
==32383==    by 0x7B373F: main (in /home/vagrant/CppMicroServices/build/bin/usDeclarativeServicesTests)
==32383==
[==========] Running 323 tests from 36 test suites.
[----------] Global test environment set-up.
[----------] 2 tests from ActivatorTest
[ RUN      ] ActivatorTest.VerifySCRService
```

File: cmake/usFunctionCheckCompilerFlags.cmake [hunk 3]
```diff
@@ -62,7 +62,13 @@ function(usFunctionCheckCompilerFlags FLAG_TO_TEST RESULT_VAR)
   CHECK_CXX_COMPILER_FLAG(${FLAG_TO_TEST} HAS_CXX_FLAG_${suffix})
 
   if(HAS_CXX_FLAG_${suffix})
-    set(${RESULT_VAR} "${_saved_result_var} ${FLAG_TO_TEST}" PARENT_SCOPE)
+    # If the variable passed in as RESULT_VARIABLE was initially empty, then do not
+    # include it in the new value for REUSLT_VAR since it adds 1 extra space
```
[the remark this answers is not in the document]
Comment (jeffdiclemente) [reply cmake/usFunctionCheckCompilerFlags.cmake:66]: still a typo: REUSLT_VAR

File: compendium/DeclarativeServices/src/metadata/ComponentMetadata.hpp [hunk 4]
```diff
@@ -60,7 +61,58 @@ struct ComponentMetadata
...
+  os << "\tconfigurationPids[size = " << metadata.configurationPids.size() << "]: [" << std::endl;
+  for (const auto& pid : metadata.configurationPids) {
+    bool isLast = pid != metadata.configurationPids.back();
+    os << "\t\t" << pid << (isLast ? "" : ",") << std::endl;
+  }
+  os << "\t" << "]" << std::endl
+     << "\t" << "factoryComponentID: " << metadata.factoryComponentID << std::endl;
+  os << "\tfactoryComponentProperties: [size = " << metadata.factoryComponentProperties.size() << "] not shown" << std::endl;
+
+  return os;
```
Comment (jeffdiclemente) [inline compendium/DeclarativeServices/src/metadata/ComponentMetadata.hpp:114]: I'd prefer if the operator<< is only available for tests. We don't need to provide an operator<< to API users. See CppMicroServices\compendium\DeclarativeServices\test\TestReferenceManagerImpl.cpp for an example.

**Review by jeffdiclemente — CHANGES_REQUESTED** [review]

**Review by jeffdiclemente — APPROVED** [review]
still a typo in cmake/usFunctionCheckCompilerFlags.cmake; REUSLT_VAR -> RESULT_VAR. once that's fixed this PR can be merged into development.

**Pull request merged**